File: 659-Split Array into Consecutive Subsequences.py
from collections import Counter
class Solution(object):
    def isPossible(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
# greedy
# consider 3 cases:
# 1. add current number to previous consecutive subseqence if exists
# 2. check if current number + 1 & current number + 2 exists, if yes, create a new consecutive subsequence
# 3. return false

        freq = Counter(nums)
        last = defaultdict(int)
        
        for num in nums:
            if freq[num]:
                if last[num-1]:
                    last[num-1]-=1
                    last[num]+=1
                elif freq[num+1] and freq[num+2]:
                    freq[num+1]-=1
                    freq[num+2]-=1
                    last[num+2]+=1
                else:
                    return False
                freq[num]-=1
        
        return True

File: JavaScript/659-Split Array into Consecutive Subsequences.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var isPossible = function(nums) {
    let freq = new Map();
    for (let num of nums){
        freq.set(num, (freq.get(num)|| 0) + 1 )
    }
    
    let last = new Map();
    
    for (let num of nums){
        if (freq.get(num)){
            if (last.has(num-1) && last.get(num-1) !==0){
                last.set(num-1, last.get(num-1)-1);
                last.set(num, (last.get(num)||0)+1);
            }else if(freq.get(num+1) && freq.get(num+2)){
                freq.set(num+1, freq.get(num+1) - 1);
                freq.set(num+2, freq.get(num+2) - 1);
                last.set(num+2, (last.get(num+2)||0) + 1);
            }else{
                return false;
            }
            freq.set(num, freq.get(num)-1);
        }
    }
    return true;
};